x/proof/keeper/msg_server_submit_proof.go

```diff
@@ -127,12 +224,141 @@ func (k msgServer) queryAndValidateClaimForProof(ctx context.Context, proof *typ
+		return types.ErrProofInvalidRelay.Wrapf(
+			"sessionHeaders session end heights mismatch expect: %d, got: %d",
+			expectedSessionHeader.GetSessionEndBlockHeight(),
+			sessionHeader.GetSessionEndBlockHeight(),
+		)
+	}
+
+	if sessionHeader.GetSessionId() != expectedSessionHeader.GetSessionId() {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"sessionHeaders session IDs mismatch expect: %q, got: %q",
+			expectedSessionHeader.GetSessionId(),
+			sessionHeader.GetSessionId(),
+		)
+	}
+
+	return nil
+}
+
+// verifyClosestProof verifies the closest merkle proof against the expected root hash.
+func verifyClosestProof(
+	proof *smt.SparseMerkleClosestProof,
+	expectedRootHash []byte,
+) error {
+	spec := smt.NoPrehashSpec(sha256.New(), true)
+
+	valid, err := smt.VerifyClosestProof(proof, expectedRootHash, spec)
+	if err != nil {
+		return err
+	}
+
+	if !valid {
+		return types.ErrProofInvalidProof.Wrap("invalid closest merkle proof")
+	}
+
+	return nil
+}
+
+// validateMiningDifficulty ensures that the relay's mining difficulty meets the required
+// difficulty.
+// TODO_TECHDEBT: Factor out the relay mining difficulty validation into a shared function
+// that can be used by both the proof and the miner packages.
+func validateMiningDifficulty(relayBz []byte) error {
+	hasher := sha256.New()
+	hasher.Write(relayBz)
+	realyHash := hasher.Sum(nil)
+
+	difficultyBits, err := protocol.CountDifficultyBits(realyHash)
+	if err != nil {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"error counting difficulty bits: %s",
+			err,
+		)
+	}
+
+	if difficultyBits < relayMinDifficultyBits {
+		return types.ErrProofInvalidRelay.Wrapf(
+			"relay difficulty %d is less than the required difficulty %d",
+			difficultyBits,
+			relayMinDifficultyBits,
+		)
+	}
+
+	return nil
+}
+
+// validateClosestPath ensures that the proof's path matches the expected path.
+func (k msgServer) validateClosestPath(
+	ctx context.Context,
+	proof *smt.SparseMerkleClosestProof,
+	sessionHeader *sessiontypes.SessionHeader,
+) error {
+	// The RelayMiner has to wait until the createClaimWindowStartHeight and the
+	// submitProofWindowStartHeight are open to respectively create the claim and
+	// submit the proof.
+	// These windows are calculated as SessionEndBlockHeight + GracePeriodBlockCount.
+	// see: relayerSessionsManager.waitForEarliest{CreateClaim,SubmitProof}Height().
+	// The RelayMiner hast to wait this long to ensure that late relays are accepted
+	// and included in the SessionNumber=N tree.
+	// (i.e. relays initiated by Applications/Gateways in SessionNumber=N but
+	// arriving to the RelayMiner in SessionNumber=N + 1)
+	// Otherwise, the RelayMiner would not be able to include the late relays in
+	// the SessionNumber N claim and the proof.
+	// Since smt.ProveClosest is in terms of submitProofWindowStartHeight, this
+	// block's hash needs to be used for validation too.
+	// TODO_TECHDEBT(#409): Reference the session rollover documentation here.
```

Comment:
```suggestion
	// The RelayMiner has to wait until the createClaimWindowStartHeight and the
	// submitProofWindowStartHeight are open to create the claim & submit the proof respectively.
	//
	// These windows are calculated as (SessionEndBlockHeight + GracePeriodBlockCount).
	// For reference, see relayerSessionsManager.waitForEarliest{CreateClaim,SubmitProof}Height().
	//
	// The RelayMiner has to wait this long to ensure that late relays (i.e. submitted during
	// SessionNumber=(N+1) but created during SessionNumber=N) are still included as part
	// of SessionNumber=N. 
	//
	// Since smt.ProveClosest is defined in terms of submitProofWindowStartHeight, this
	// block's hash needs to be used for validation too.
	//
	// TODO_TECHDEBT(#409): Reference the session rollover documentation here.
```